### PR TITLE
Create: 0093-restore-ip-addresses.js

### PR DESCRIPTION
**Pull request by Sbwal**
[//]: # "Pull Request Template"
[//]: # "Replace the placeholder values in the template below"

- **File(s) Modified**: _0093-restore-ip-addresses.js_
- **Language(s) Used**: _javascript_
- **Submission URL**: _https://leetcode.com/problems/restore-ip-addresses/submissions/882520828/_

[//]: # "Getting the Submission URL"
[//]: # "Go to the leetcode [`Submissions tab`](https://user-images.githubusercontent.com/71089234/180188604-b1ecaf90-bf27-4fd6-a559-5567aebf8930.png)"
[//]: # "and [click on the `Accepted` status of your submission.](https://user-images.githubusercontent.com/71089234/180189321-1a48c33f-aa65-4b29-8aaa-685f4f5f8c9e.png)]"
[//]: # "Finally copy the URL from the nav bar, it should look like https://leetcode.com/problems/[problem-name]/submissions/xxxxxxxxx/"

### Important
Please make sure the file name is lowercase and a duplicate file does not already exist before merging.

### CODE REVIEW

**Comment by aakhtar3** [comment]
I'm getting a "Submission does not exist" from your link.

File: javascript/0093-restore-ip-addresses.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+            return;
+        }
+        if(d > 4) return;
+        for(let j = i; j < Math.min(i + 3, s.length); j++){
+            let p = s.substring(i, j + 1)
+            if(Number(p) > 255 || (p.length > 1 && p[0] === '0')) continue;
+            bt(j + 1, d + 1, ip + p + '.');
+        }
+    }
+    bt(0, 0, '');
```
Comment (aakhtar3) [inline javascript/0093-restore-ip-addresses.js:20]: I'm not seeing a clearing backtracking solution, but just a normal DFS implementation. Take a peek here.

```js
    /**
 * DFS - BackTrack
 * Time O((M^N) * N) | Space O(M * N)
 * https://leetcode.com/problems/restore-ip-addresses
 * @param {string} s
 * @return {string[]}
 */
var restoreIpAddresses = (s, start = 0, dots = [], ips = []) => {
    const [ length, numbers ] = [ (s.length - start), (4 - dots.length) ];

    if (isBaseCase1(length, numbers)) return ips;
    if (isBaseCase2(s, start, dots, length)) return ips;

    const canRestore = (dots.length === 3);
    if (canRestore) restore(s, start, dots, ips);

    backTrack(s, start, dots, ips, length)

    return ips;
}

const isBaseCase1 = (length, numbers) => {
    const isInRange = ((numbers * 3) < length);
    const isOutOfRange = (length < numbers);

    return (isInRange || isOutOfRange);
}

const isBaseCase2 = (s, start, dots, length) => {
    const isOctet = (dots.length === 3);
    
    return (isOctet && !isIp(s, start, length))
}

var isIp = (s, start, length) => {
    const isOne = (length === 1);
    if (isOne) return isOne;

    const hasLeadingZero = (s[start] === '0');
    const isOctet = (length < 3);
    const isInRange = (s.slice(start, (start + length)).localeCompare("255") <= 0);

    return (!hasLeadingZero && (isOctet || isInRange));
}

var restore = (s, start, dots, ips, last = 0, sb = []) => {
    for (const dot of dots) {
        sb.push(s.slice(last, (last + dot)));
        sb.push('.');
        last += dot;
    }

    sb.push(s.slice(start));
    ips.push(sb.join(''));
}

const backTrack = (s, start, dots, ips, length) => {
    for (let end = 1; (end <= 3) && (end <= length); end++) {        
        dots.push(end);
            if (isIp(s, start, end)) restoreIpAddresses(s, (start + end), dots, ips);
        dots.pop();
    }
}

/**
 * Iterative
 * Time O((M^N) * N) | Space O(M * N)
 * https://leetcode.com/problems/restore-ip-addresses
 * @param {string} s
 * @return {string[]}
 */
var restoreIpAddresses = (s, ips = []) => {
    let len1 = Math.max(1, (s.length - 9));
    const isLen1InRange = (s, len1) => ((len1 <= 3) && (len1 <= (s.length - 3)));

    while (isLen1InRange(s, len1)) {
        if (!isIp(s, 0, len1)) { len1++; continue; }

        let len2 = Math.max(1, ((s.length - len1) - 6));
        const isLen2InRange = (s, len1, len2) => ((len2 <= 3) && (len2 <= ((s.length - len1) - 2)));

        while (isLen2InRange(s, len1, len2)) {
            if (!isIp(s, len1, len2)) { len2++; continue; }

            let len3 = Math.max(1, (((s.length - len1) - len2) - 3));
            const isLen3InRange = (s, len1, len2, len3) => ((len3 <= 3) && (len3 <= (((s.length - len1) - len2) - 1)));

            while (isLen3InRange(s, len1, len2, len3)) {
                const isValid = (isIp(s, (len1 + len2), len3) && isIp(s, ((len1 + len2) + len3),  (((s.length - len1) - len2) - len3)))
                if (!isValid) { len3++; continue; }

                restore(s, len1, len2, len3, ips);
                len3++;
            }

            len2++;
        }

        len1++;
    }

    return ips;
}


var isIp = (s, start, length) => {
    const isOne = (length === 1);
    if (isOne) return isOne;

    const hasLeadingZero = (s[start] === '0');
    const isOctet = (length < 3);
    const isInRange = (s.slice(start, (start + length)).localeCompare("255") <= 0);

    return (!hasLeadingZero && (isOctet || isInRange));
}

var restore = (s, len1, len2, len3, ips) => {
    const sb = [
        s.slice(0, len1),
        s.slice(len1, len1 + len2),
        s.slice((len1 + len2), ((len1 + len2) + len3)),
        s.slice(((len1 + len2) + len3))
    ];

    ips.push(sb.join('.'));
}

```

Comment (Sbwal) [reply javascript/0093-restore-ip-addresses.js:20]: image
Hi I have shared the correct link. Please check again.
Also, the code is same as that explained in video using DFS approach. Before this, I have submitted another solution 
`/**
 * @param {string} s
 * @return {string[]}
 */
var restoreIpAddresses = function(s) {
    const a = [], ip = [];
    const bt = (i, p) => {
        if(i === s.length && ip.length === 4) a.push(ip.join('.'));
        if(i === s.length || ip.length === 4) return;
        if((p.length > 0 && p[0] === '0') || Number(p + s[i]) > 255) return;
        ip.push(p + s[i]);
        bt(i + 1, '');
        ip.pop();
        bt(i + 1, p + s[i]);
    }
    bt(0, '');
    return a;
};`
Submission Link - https://leetcode.com/problems/restore-ip-addresses/submissions/882392338/
Here, it's same as you are expecting. However, when people while go through the video and look for solution that is not explained than they **might** get confused.

**Comment by Ykhan799** [comment]
Backtracking solution already exists

**Pull request closed without merging**